### PR TITLE
[CLI] - Add commands for config set-all and rm-all

### PR DESCRIPTION
**Pull request by komalali**
Fixes: https://github.com/pulumi/pulumi/issues/6329

### CODE REVIEW

**Review by justinvp — APPROVED** [review]
LGTM

**Review by EvanBoyle — COMMENTED** [review]
One edge case for parsing that we should add test cases for at a minimum.

That sort of leads me to the question of whether or not we should be using the formal:
`--plaintext k=v --secret plain=v`

You could imagine it also looking like:
`k v k1 v2 --secret k3 v3 k4 v4 --secret k5 v5`

But this would require setting `DisableFlagParsing` and then manually handling the flags via the args function. Definitely reasons not to do this, and not sure if it would be considered idiomatic or not. Requiring `--plaintext` is a little annoying, but the cases we have in mind are more machine-oriented so I'm a little less concerned about it. 

I could really see things going either way, curious to hear your thoughts.

**Review by stack72 — COMMENTED** [review]
My only feedback here is that the use of `--plaintext` feels like it's a bit more overhead for users - I'm not disagreeing with the implementation but just thinking from a user perspective

File: pkg/cmd/pulumi/config.go [hunk 1]
```diff
@@ -326,6 +328,56 @@ func newConfigRmCmd(stack *string) *cobra.Command {
...
+func newConfigRmAllCmd(stack *string) *cobra.Command {
+	var path bool
+
+	rmAllCmd := &cobra.Command{
+		Use:   "rm-all <key1> <key2> <key3> ...",
+		Short: "Remove multiple configuration values",
+		Long: "Remove multiple configuration values.\n\n" +
+			"The `--path` flag indicates that keys should be parsed within maps or lists:\n\n" +
+			"  - `pulumi config rm-all outer.inner foo[0] key1` will remove the " +
+			"`outer.inner`, `foo[0]` and `key1` keys\n" +
```
[the remark this answers is not in the document]
Comment (komalali) [reply pkg/cmd/pulumi/config.go:340]: Yes. There's a little more context over in the [issue](https://github.com/pulumi/pulumi/issues/6329#issuecomment-778516062) around how we got here. If you have ideas on how to make this clearer, let me know.

**Comment by komalali** [comment]
Fair feedback @stack72 and @EvanBoyle. 

The reason behind enforcing `--plaintext` has to do with making the user explicitly acknowledge that they are **not** storing this value as a secret, but I agree it is a little kludgey and removing it doesn't necessarily complicate things so I could be convinced. The question is do we value being explicit about secret vs non-secret more than the burden of having to type `--plaintext` every time 🤷 - I could be convinced either way.

The edge case that @EvanBoyle brings up (where there is a `=` in the key) is a little grosser to manage. We could certainly go with what you're suggesting, (i.e. `k v k1 v2 --secret k3 v3 k4 v4 --secret k5 v5`) but it certainly complicates the implementation. Let me try to work through the edge case with the current `key=value` structure and if that proves impossible we can consider the cost of the complicated implementation vs the likelihood of a key including `=`.

**Comment by stack72** [comment]
> Fair feedback @stack72 and @EvanBoyle.
> 
> The reason behind enforcing `--plaintext` has to do with making the user explicitly acknowledge that they are **not** storing this value as a secret (since we aren't doing any interactive checking to see if the value looks like a secret or not), but I agree it is a little kludgey and removing it doesn't necessarily complicate things so I could be convinced. The question is do we value being explicit about secret vs non-secret more than the burden of having to type `--plaintext` every time 🤷 - I could be convinced either way.
> 
> The edge case that @EvanBoyle brings up (where there is a `=` in the key) is a little grosser to manage. We could certainly go with what you're suggesting, (i.e. `k v k1 v2 --secret k3 v3 k4 v4 --secret k5 v5`) but it certainly complicates the implementation. Let me try to work through the edge case with the current `key=value` structure and if that proves impossible we can consider the cost of the complicated implementation vs the likelihood of a key including `=`.

Totally not something I am going to die on either :) I was just thinking about the existing functionality of config set but this has a slightly different use case

File: pkg/cmd/pulumi/config.go [hunk 2]
```diff
@@ -326,6 +328,56 @@ func newConfigRmCmd(stack *string) *cobra.Command {
...
+					return errors.Wrap(err, "invalid configuration key")
+				}
+
+				err = ps.Config.Remove(key, path)
+				if err != nil {
+					return err
+				}
+			}
+
+			return saveProjectStack(s, ps)
```
[the remark this answers is not in the document]
Comment (komalali) [reply pkg/cmd/pulumi/config.go:371]: There's a network call every time there's a secret encrypted or decrypted... but other than that it seems that the network call only happens when there's an `up`.

[the remark this answers is not in the document]
Comment (EvanBoyle) [reply pkg/cmd/pulumi/config.go:371]: I read through the code again, and I don't think this is going to help with the underlying latency issues. Secrets are the bottle-neck here it seems and this implementation loops through encrypting values one at a time for `set-all --secret`, which should be more or less equivalent to running `pulumi config set --secret` sequentially. We might need to think about spawing a go routine to do this in parallel, or creating a bulk config encryption endpoint in the service if one doesn't already exist.

It would be helpful to do some benchmarking here to understand performance impacts for a few different scenarios (bulk set plaintext, bulk set a bunch of secrets, etc).

In addition, the other underlying issue that I thought we were tackling with this, but didn't see mentioned in https://github.com/pulumi/pulumi/issues/6329 was the parallelism issues with setting config. Running config command async in a loop across stacks was causing things to get garbled because stack selection was changing. Does this change help with that? I notice that these two methods use `setCurrent true` when getting the stack so I'm not sure. Would probably be good to test as well.

[the remark this answers is not in the document]
Comment (komalali) [reply pkg/cmd/pulumi/config.go:371]: I know with certainty that this greatly improves the latency experienced by automation api: https://github.com/pulumi/pulumi/issues/6042#issuecomment-781807632
But yes, I'll consider the go routine as an additional improvement if this isn't satisfactory.

> I notice that these two methods use setCurrent true when getting the stack so I'm not sure.
That's a mistake, it should be `false`. In fact, `set-all` is false, I just made an error here.

[the remark this answers is not in the document]
Comment (komalali) [reply pkg/cmd/pulumi/config.go:371]: Okay, since my other profiling test was in python, I did some in Go this time.

Test: setting 20 config keys
Before:
* Mixed (10 secret, 10 plain): 11.69s
* Plain (20 plain): 11.25s
* Secret (20 secret): 12.29s

After:
* Mixed (10 secret, 10 plain): 1.18s
* Plain (20 plain): 1.05s
* Secret (20 secret): 1.51s

I think the results speak for themselves.

File: pkg/cmd/pulumi/config.go [hunk 3]
```diff
@@ -507,6 +559,108 @@ func newConfigSetCmd(stack *string) *cobra.Command {
...
+	setCmd.PersistentFlags().StringArrayVar(
+		&secretArgs, "secret", []string{},
+		"Marks a value as secret to be encrypted")
+
+	return setCmd
+}
+
+func parseKeyValuePair(pair string) (config.Key, string, error) {
+	// Split the arg on the first '=' to separate key and value.
+	splitArg := strings.SplitN(pair, "=", 2)
```
[the remark this answers is not in the document]
Comment (komalali) [reply pkg/cmd/pulumi/config.go:651]: Fixed

**Review by EvanBoyle — APPROVED** [review]
🚀

File: pkg/cmd/pulumi/config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (EvanBoyle) [reply pkg/cmd/pulumi/config.go:371]: Sorry I missed that profiling data in the linked issue. Glad to hear we've got this sorted out!
